tsdb/engine/tsm1/data_file.go

```diff
@@ -0,0 +1,666 @@
+}
+
+func (d *indirectIndex) readKey(b []byte) (n int, key string, err error) {
+	// 2 byte size of key
+	n, size := 2, int(btou16(b[:2]))
+
+	// N byte key
+	key = string(b[n : n+size])
+	n += len(key)
+	return
+}
+
+func (d *indirectIndex) readEntries(b []byte) (n int, entries indexEntries, err error) {
+	// 2 byte count of index entries
+	n, count := 2, int(btou16(b[:2]))
+
+	for i := 0; i < count; i++ {
+		ie := &IndexEntry{}
+		if err := ie.UnmarshalBinary(b[i*indexEntrySize+2 : i*indexEntrySize+2+indexEntrySize]); err != nil {
+			return 0, nil, fmt.Errorf("readEntries: unmarshal error: %v", err)
+		}
+		entries = append(entries, ie)
+		n += indexEntrySize
+	}
+	return
+}
+
+// tsmWriter writes keys and values in the TSM format
+type tsmWriter struct {
+	w     io.Writer
+	index TSMIndex
+	n     int64
+}
+
+func NewTSMWriter(w io.Writer) (TSMWriter, error) {
+	n, err := w.Write(append(u32tob(MagicNumber), Version))
+	if err != nil {
+		return nil, err
+	}
+
+	index := &directIndex{
+		blocks: map[string]indexEntries{},
+	}
+
+	return &tsmWriter{w: w, index: index, n: int64(n)}, nil
+}
+
+func (t *tsmWriter) Write(key string, values Values) error {
+	block, err := values.Encode(nil)
+	if err != nil {
+		return err
+	}
+
+	checksum := crc32.ChecksumIEEE(block)
+
+	n, err := t.w.Write(append(u32tob(checksum), block...))
+	if err != nil {
+		return err
+	}
+
+	// Record this block in index
+	t.index.Add(key, values[0].Time(), values[len(values)-1].Time(), t.n, uint32(n))
+
+	// Increment file position pointer
+	t.n += int64(n)
+	return nil
+}
+
+func (t *tsmWriter) Close() error {
+	indexPos := t.n
+
+	// Generate the index bytes
+	b, err := t.index.MarshalBinary()
+	if err != nil {
+		return err
+	}
+
+	// Write the index followed by index position
+	_, err = t.w.Write(append(b, u64tob(uint64(indexPos))...))
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+type tsmReader struct {
+	r                    io.ReadSeeker
+	indexStart, indexEnd int64
+	index                TSMIndex
+}
+
+func NewTSMReader(r io.ReadSeeker) (*tsmReader, error) {
+	t := &tsmReader{r: r}
+	if err := t.init(); err != nil {
+		return nil, err
+	}
+
+	return t, nil
+}
+
+func (t *tsmReader) init() error {
+	// Current the readers size
+	size, err := t.r.Seek(0, os.SEEK_END)
+	if err != nil {
+		return fmt.Errorf("init: failed to seek: %v", err)
+	}
+
+	t.indexEnd = size - 8
+
+	// Seek to index location pointer
+	_, err = t.r.Seek(-8, os.SEEK_END)
+	if err != nil {
+		return fmt.Errorf("init: failed to seek to index ptr: %v", err)
+	}
+
+	// Read the absolute position of the start of the index
+	b := make([]byte, 8)
+	_, err = t.r.Read(b)
+	if err != nil {
+		return fmt.Errorf("init: failed to read index ptr: %v", err)
+
+	}
+
+	t.indexStart = int64(btou64(b))
+
+	_, err = t.r.Seek(t.indexStart, os.SEEK_SET)
+	if err != nil {
+		return fmt.Errorf("init: failed to seek to index: %v", err)
+	}
+
+	b = make([]byte, t.indexEnd-t.indexStart)
+	t.index = &directIndex{
```

Comment:
For this PR yes.  I'll be updating this in subsequent PRs when I bring back the MMAP version.